### ISSUE TITLE
Experiments as input to indexing

### ISSUE THREAD

**Issue by rjgildea** [dials/dials#545]
In Berkeley we had a discussion about the problem of how to handle indexing of multiple datasets (e.g. multiple still images), and how that was affected by the choice of imageset data structures, e.g. differences between imagesweeps and imagesets. Having thought about this some more, I believe that what indexing really needs as input is a set of crystal-less experiments, to which indexing can then attach a set of crystal models (or a shared crystal model across multiple experiments in the case of joint-indexing). Indexing doesn't actually need/care about the underlying imagesets, as it doesn't need to read image data, needing only the experiment models (detector, beam, scan, goniometer, crystal).

I suggest modifications to the import step (and probably underlying dxtbx experiments/datablocks) to setup the appropriate crystal-less experiments. There are a number of scenarios that would need to be handled (please comment if there are any I've missed here):

1. Single rotation dataset
- input: multiple images with shared template (or multi-image file)
- output: single experiment

2. Single still image
- input: single image file
- output: single experiment

3. Multiple rotation datasets (this could be narrow or wide sweeps, or even single rotation images)
- input: multiple images with different shared templates (or multi-image file)
- output: multiple experiments
- possible flag at import stage to say whether multiple datasets correspond to same or different samples (e.g. for joint indexing/refinement)

4. Multiple still images
- input: multiple images (shared or unrelated template, or multi-image file)
- output: multiple experiments
- possible flag at import stage to say whether multiple datasets correspond to same or different samples, or is gridscan, etc. (alternatively a "grid scan" might be a special type of `scan` model)

This would have the added benefit of simplifying things for users, without having to understand the subtly confusing difference between datablock.json and experiments.json, as we would now just have a single format, experiments.json, that is carried all the way through the processing. Each step in the process simply adds to, or modifies information in, the experimental models.

Alongside this, we might also need new tools to allow the user to modify the experiments.json, e.g. to specify sharing of various models between experiments, or conversely to convert shared models into independent models (e.g. a user might want shared detector models at the indexing stage, but then independent detector models for later refinement after crystal symmetry has been enforced).

I would be interested in hearing feedback on these suggestions.

**Comment by graeme-winter** [comment]
Possibly the approach I would take with this would be to sketch out a new command-line tool ```dials.create_experiment``` which can be developed in parallel with ```dials.import``` so nothing is changed in terms of the typical work flow, until we move over. 

I can completely see the benefit of this - the experiment model would then run from cradle to grave for the data (i.e. import images to exporting scaled data)

**Comment by graeme-winter** [comment]
Would also have benefit of having a place where we can explicitly construct the experiments rather than the current implicit / guess set up to get to:
![gm5043fig4](https://user-images.githubusercontent.com/1241716/39233582-ab8bd4ea-4868-11e8-9d57-f2c8de59125b.jpg)
 - fun bit will be working out how to define a UI to this which is straightforward for the most common use cases...

**Comment by dagewa** [comment]
My feeling is that the proposal makes a lot of sense, and ticks quite a few boxes beyond the initial issue, all of which are useful changes. For example, moving away from the datablock structure by allowing incomplete experiments (i.e. missing crystal) simplifies the framework. We currently have some tools that have to handle both data structures, and some cases where the tool is essentially duplicated (see `dials.combine_experiments` and `dev.dials.combine_datablocks`). We really need this comprehensive set of tools for creating, modifying and combining multiple experiments for multi-crystal work. Although DIALS has facilities for handling multiple experiments, in practice they are not particularly user-friendly. Revisiting this from the ground up, by cleaning up the data structures, seems to me like the right approach.

**Comment by jmp1985** [comment]
I agree with what's been said above. Removing the datablock class has been on the dxtbx refactor list for quite a while. Having a subtly different file format for metadata without crystal model has always been confusing, so I think it would definitely be a good idea to remove in favour of just keeping the experiment list.

I also think it would be nice to have a full suite of well designed tools for modifying and creating experiment files. dials.combined/split_experiments do the job but aren't very user friendly. I guess we could have commands something like this using the new experiment identifier functionality. 

```
#  Create the initial experiments.json file
dials.experiments.create  /path/to/images/1/  id="apple"

# Add experiments to the experiments file
dials.experiments.add      /path/to/images/2/  id="orange"  experiments.json
dials.experiments.add      /path/to/images/2/  id="banana"  experiments.json

# Specify shared models
dials.experiments.link experiments.json \
  detector.reference="apples" \
  detector.linked="oranges" \
  detector.linked="banana" \

# Remove experiments from a file
dials.experiments.remove experiments.json id="banana"

# Remove links between models
dials.experiments.unlink experiments.json \
  detector.unlink="oranges" 
```

**Comment by dagewa** [comment]
An interactive mode does sound useful. For large numbers of experiments I have imagined a GUI with interactive graph diagrams like the above could be a useful feature, potentially with clustering of similar models.

**Comment by rjgildea** [comment]
It would appear from the existing tests in dxtbx that "crystal-less" experiments are supported and are tested:
https://github.com/cctbx/cctbx_project/blob/master/dxtbx/tests/model/experiment/test_experiment_list.py#L292
